beluga_tutorial/src/main.cpp

```diff
@@ -0,0 +1,278 @@
+    estimation["mean"] = std::get<0>(data.estimation);
+    estimation["sd"] = std::get<1>(data.estimation);
+  }
+
+  std::filesystem::create_directory(path.parent_path());
+  std::ofstream fout(path, std::ios::trunc);
+  fout << node;
+  fout.close();
+}
+
+int main(int argc, char* argv[]) {
+  if (argc < 3) {
+    std::cerr << "Usage: " << argv[0] << "<params_directory_path> <params_filename>\n";
+    return 1;
+  }
+
+  const std::string directory = argv[1];
+  const std::string filename = argv[2];
+  const auto parameters_file_path = std::filesystem::path{directory} / filename;
+
+  TutorialParams tutorial_params;
+  load_params_from_yaml(parameters_file_path, tutorial_params);
+
+  // Create the TutorialDataset object for record the simulation data
+  TutorialDataset tutorial_dataset;
+
+  LandmarkMapVector landmark_map;
+  load_landmark_map_from_yaml(std::filesystem::path{directory} / tutorial_params.landmark_map_file_name, landmark_map);
+  tutorial_dataset.landmark_map = landmark_map;
+
+  // Generate particles
+  std::normal_distribution<double> initial_distribution(tutorial_params.initial_pose, tutorial_params.initial_pose_sd);
+  auto particles = beluga::views::sample(initial_distribution) |                       //
+                   ranges::views::transform(beluga::make_from_state<Particle>) |       //
+                   ranges::views::take_exactly(tutorial_params.number_of_particles) |  //
+                   ranges::to<beluga::TupleVector>;
+
+  // Execute the particle filter using beluga
+  double current_pose{tutorial_params.initial_pose};
+  for (auto n = 0; n < tutorial_params.number_of_cycles; ++n) {
+    TutorialData tutorial_data;
+
+    // Save ground truth
+    current_pose += tutorial_params.velocity * tutorial_params.dt;
+    tutorial_data.ground_truth = current_pose;
+
+    // Check if the simulation is out of bounds
+    if (current_pose > tutorial_params.map_size) {
+      break;
+    }
+
+    auto motion_model = [&](double state, auto& gen) {
+      const double distance = (tutorial_params.velocity * tutorial_params.dt);
+      std::normal_distribution<double> distribution(distance, tutorial_params.translation_sd);
+      return state + distribution(gen);
+    };
+
+    // Generate simulated sensor data
+    auto sensor_data =
+        landmark_map |                                                                                            //
+        ranges::views::transform([&](double landmark) { return landmark - current_pose; }) |                      //
+        ranges::views::remove_if([&](double range) { return std::abs(range) > tutorial_params.sensor_range; }) |  //
+        ranges::to<SensorData>;
+
+    auto sensor_model = [&](const double& state) {
+      auto particle_sensor_data = landmark_map |                                                           //
+                                  ranges::views::transform([&](const double lm) { return lm - state; }) |  //
```

Comment:
```suggestion
                                  ranges::views::transform([&](double lm) { return lm - state; }) |  //
```

Reply:
Just to understand, why do you suggest modifying `lm` to be a non-const argument?